### ISSUE TITLE
Camera preview has channels reversed on Android hardware (or emulator)

### ISSUE THREAD

**Issue by hoonkai** [opencv/opencv#11118]
##### System information (version)

- OpenCV => 3.4.1
- Operating System / Platform => Linux Mint 18.1 64-bit
- Compiler => GCC 5.4.0
- Android API => 24

##### Detailed description

Some straightforward code on Android gives rise to opposite colour channels on emulator and hardware. When `onCameraFrame(CvCameraViewFrame inputFrame)` is override and `inputFrame` is converted to as BGR, emulator and hardware show opposite results.

Emulator:
<img src=https://user-images.githubusercontent.com/14197204/37613533-fa80f92c-2bf3-11e8-86d4-49e40b575e0e.png width=200/>
Hardware (Galaxy S7):
<img src=https://user-images.githubusercontent.com/14197204/37613551-0547c138-2bf4-11e8-89fa-642e34e02ff4.png width=200/>

##### Steps to reproduce

Example program for reproduction:

```java
public class OpenCVCamera extends AppCompatActivity implements CameraBridgeViewBase.CvCameraViewListener2 {
    private static final String TAG = "OpenCVCamera";

    private CameraBridgeViewBase cameraBridgeViewBase;

    private Mat mRgba;

    private int requestCode;
    private int grantResults[];


    private BaseLoaderCallback baseLoaderCallback = new BaseLoaderCallback(this) {
        @Override
        public void onManagerConnected(int status) {
            switch (status) {
                case LoaderCallbackInterface.SUCCESS:
                    cameraBridgeViewBase.enableView();
                    break;
                default:
                    super.onManagerConnected(status);
                    break;
            }
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getSupportActionBar().hide();

        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        setContentView(R.layout.activity_opencv_camera);
        cameraBridgeViewBase = (CameraBridgeViewBase) findViewById(R.id.camera_view);
        cameraBridgeViewBase.setVisibility(SurfaceView.VISIBLE);
        cameraBridgeViewBase.setCvCameraViewListener(this);

        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.CAMERA)
                != PackageManager.PERMISSION_GRANTED) {

            requestCameraPermission();

        }

        if(ContextCompat.checkSelfPermission(this,Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_DENIED ) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, requestCode);
            onRequestPermissionsResult(requestCode, new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, grantResults);
        }

    }

    @Override
    public void onPause() {
        super.onPause();
        if (cameraBridgeViewBase != null)
            cameraBridgeViewBase.disableView();
    }

    @Override
    public void onResume(){
        super.onResume();
        if (!OpenCVLoader.initDebug()) {
            Log.d(TAG, "Internal OpenCV library not found. Using OpenCV Manager for initialization");
            OpenCVLoader.initAsync(OpenCVLoader.OPENCV_VERSION_3_1_0, this, baseLoaderCallback);
        } else {
            Log.d(TAG, "OpenCV library found inside package. Using it!");
            baseLoaderCallback.onManagerConnected(LoaderCallbackInterface.SUCCESS);
        }
    }

    public void onDestroy() {
        super.onDestroy();
        if (cameraBridgeViewBase != null)
            cameraBridgeViewBase.disableView();
    }


    @Override
    public void onCameraViewStarted(int width, int height) {
        mRgba = new Mat(height, width, CvType.CV_8UC4);
    }

    @Override
    public void onCameraViewStopped() {

    }


    @Override
    public Mat onCameraFrame(CameraBridgeViewBase.CvCameraViewFrame inputFrame) {
        mRgba = inputFrame.rgba();
        Imgproc.cvtColor(mRgba, mRgba, Imgproc.COLOR_RGBA2BGR);

        return mRgba;
    }

    /**
     * Requests the Camera permission.
     * If the permission has been denied previously, a SnackBar will prompt the user to grant the
     * permission, otherwise it is requested directly.
     */
    private void requestCameraPermission() {
        Log.i(TAG, "CAMERA permission has NOT been granted. Requesting permission.");

        ActivityCompat.requestPermissions(this,
                new String[]{Manifest.permission.CAMERA},
                1);
    }

    @Override // android recommended class to handle permissions
    public void onRequestPermissionsResult(int requestCode,
                                           String permissions[], int[] grantResults) {
        switch (requestCode) {
            case 1: {
                if (grantResults.length > 0
                        && grantResults[0] == PackageManager.PERMISSION_GRANTED) {

                    Log.d("permission", "granted");
                } else {
                    Toast.makeText(OpenCVCamera.this, "Permission denied to read your External storage", Toast.LENGTH_SHORT).show();
                    onDestroy();
                }
                return;
            }

        }
    }
}
```

**Comment by alalek** [comment]
> Imgproc.cvtColor(mRgba, mRgba, Imgproc.COLOR_RGBA2BGR);

"Problem" is here (not a problem actually).

**Comment by alalek** [comment]
Take a look on this [SO question](https://stackoverflow.com/questions/27083097/opencv-android-color-issue-using-camerabridgeviewbase). There are some hints for debugging.
Additionally you could take a look on [Camera2 API](https://github.com/opencv/opencv/blob/3.4.1/modules/java/generator/android-21/java/org/opencv/android/JavaCamera2View.java) (requires Android 21 / 5.0).

**Comment by hoonkai** [comment]
@alalek But what I don't understand is, even though the phone uses NV21 and the emulator uses YV12, don't these lines in `JavaCameraFrame.rgba()` normalise them? Isn't the output in RGB(A) regardless of the preview format?

``` java
        @Override
        public Mat rgba() {
            if (mPreviewFormat == ImageFormat.NV21)
                Imgproc.cvtColor(mYuvFrameData, mRgba, Imgproc.COLOR_YUV2RGBA_NV21, 4);
            else if (mPreviewFormat == ImageFormat.YV12)
                Imgproc.cvtColor(mYuvFrameData, mRgba, Imgproc.COLOR_YUV2RGB_I420, 4);  // COLOR_YUV2RGBA_YV12 produces inverted colors
            else
                throw new IllegalArgumentException("Preview Format can be NV21 or YV12");

            return mRgba;
        }
```

https://github.com/opencv/opencv/blob/master/modules/java/generator/android/java/org/opencv/android/JavaCameraView.java#L321-L326

You suggested that my line `Imgproc.cvtColor(mRgba, mRgba, Imgproc.COLOR_RGBA2BGR);` caused the 'problem', but without the conversion, the emulator gives BGR whilst the hardware gives RGB though.

**Comment by hoonkai** [comment]
@alalek But without the line `Imgproc.cvtColor(mRgba, mRgba, Imgproc.COLOR_RGBA2BGR);` the emulator shows BGR whilst the phone shows RGB. The emulator (YV12) and the phone (NV21) give opposite results. Don't these lines give normalised results though?

```java
        @Override
        public Mat rgba() {
            if (mPreviewFormat == ImageFormat.NV21)
                Imgproc.cvtColor(mYuvFrameData, mRgba, Imgproc.COLOR_YUV2RGBA_NV21, 4);
            else if (mPreviewFormat == ImageFormat.YV12)
                Imgproc.cvtColor(mYuvFrameData, mRgba, Imgproc.COLOR_YUV2RGB_I420, 4); // COLOR_YUV2RGBA_YV12 produces inverted colors
            else if (mPreviewFormat == ImageFormat.YUV_420_888) {
                assert (mUVFrameData != null);
                Imgproc.cvtColorTwoPlane(mYuvFrameData, mUVFrameData, mRgba, Imgproc.COLOR_YUV2RGBA_NV21);
            } else
                throw new IllegalArgumentException("Preview Format can be NV21 or YV12");

            return mRgba;
        }
```
https://github.com/opencv/opencv/blob/master/modules/java/generator/android-21/java/org/opencv/android/JavaCamera2View.java#L332-L344

**Comment by hoonkai** [comment]
I should add that if https://github.com/opencv/opencv/blob/master/modules/java/generator/android-21/java/org/opencv/android/JavaCamera2View.java#L336 was changed from

```Imgproc.cvtColor(mYuvFrameData, mRgba, Imgproc.COLOR_YUV2RGB_I420, 4);```

to

```Imgproc.cvtColor(mYuvFrameData, mRgba, Imgproc.COLOR_YUV2RGBA_YV12, 4);```

RGB is shown in emulator.

**Comment by alalek** [comment]
AFAIK, NV21 is a default format. And support of this format is required by Android OS (to be conformant to Android API).

https://developer.android.com/reference/android/hardware/Camera.PreviewCallback.html:
> If setPreviewFormat(int) is never called, the default will be the YCbCr_420_SP (NV21) format.

Please debug [this part of code](https://github.com/opencv/opencv/blob/3.4.1/modules/java/generator/android/java/org/opencv/android/JavaCameraView.java#L149-L162) on your device.

**Comment by hoonkai** [comment]
@alalek Yes, I think the suggested part of the code gives expected results. On the emulator:

```Build.FINGERPRINT: google/sdk_google_phone_x86/generic_x86:7.1.1/NYC/4316688:user/release-keys
Build.MODEL: Android SDK built for x86```

On the phone:

```Build.FINGERPRINT: samsung/hero2ltexx/hero2lte:7.0/NRD90M/G935FXXU1DQG1:user/release-keys
Build.MODEL: SM-G935F```

So the emulator is set to YV12 and the phone is set to NV21. But then, is `COLOR_YUV2RGB_I420` the right target format for YV12?

**Comment by hoonkai** [comment]
I've now tried the following `onCameraFrame` and the emulator shows BGR:

```java

    @Override
    public Mat onCameraFrame(CameraBridgeViewBase.CvCameraViewFrame inputFrame) {
        Imgproc.cvtColor(((JavaCameraView.JavaCameraFrame) inputFrame).mYuvFrameData, mRgba, Imgproc.COLOR_YUV2RGB_I420, 4);  // COLOR_YUV2RGBA_YV12 produces inverted colors

        return mRgba;
    }
```

where `JavaCameraFrame` and `mYuvFrameData` were made public.

**Comment by alalek** [comment]
What are values of "mPreviewFormat" on emulator / device?

**Comment by hoonkai** [comment]
Emulator (Nexus 5 API: 25) gives 0x32315659 (YV12) and device (Galaxy S7 API 24) gives 0x11 (default?).

**Comment by alalek** [comment]
`0x11` is [`NV12`](https://developer.android.com/reference/android/graphics/ImageFormat.html#NV21) (default).
`mPreviewFormat` value looks valid for both cases.

Does device work without `Imgproc.cvtColor` call (via normal `return inputFrame.rgba()` code)?
Emulator?

/cc @jmgomezpoveda (author of the latest related patch for this: #8168)

**Comment by hoonkai** [comment]
Yes, device works without `Imgproc.cvtColor` in `onCameraFrame()` (i.e., `inputFrame.rgba()` returns RGB). By the look of things, the problem seems to be `Imgproc.cvtColor(mYuvFrameData, mRgba, Imgproc.COLOR_YUV2RGB_I420, 4);`  for YV12 (https://github.com/opencv/opencv/blob/master/modules/java/generator/android-21/java/org/opencv/android/JavaCamera2View.java#L336). Like I mentioned earlier, changing `COLOR_YUV2RGB_I420` to `COLOR_YUV2RGB_YV12` gives RGB in emulator.

**Comment by hoonkai** [comment]
I believe @jmgomezpoveda  got the code here: https://stackoverflow.com/a/27233595/5349476

**Comment by hoonkai** [comment]
I think we should consider labelling this as a bug.

**Comment by tipsmiller** [comment]
I was poking around at this today. For my device/settings, frames were coming in to JavaCamera2View as `ImageFormat.YUV_420_888`. I noticed red/blue channels were switched, so after some sniffing I found that replacing the `Imgproc.COLOR_YUV2RGBA_NV21` value with `Imgproc.COLOR_YUV420sp2BGR` on [line 339](https://github.com/opencv/opencv/blob/d5afa8943b094beef09d3ae6f82d09555b5e6545/modules/java/generator/android-21/java/org/opencv/android/JavaCamera2View.java#L339) resulted in frames being delivered in the correct color space.

**Comment by ansanders** [comment]
@tipsmiller man that did the trick. I noticed the same behaviour on my Galaxy S8 and Nexus 7 (2013) 
Indeed seems like a bug in native color conversion

**Comment by tipsmiller** [comment]
I also noticed later that saving images using imwrite required converting from rgb back to bgr, otherwise the channels were once again reversed. I'm not sure what this means, perhaps android's display channels are flipped?

**Comment by tipsmiller** [comment]
Update to previous: This was due to saving images in PNG format, which expects BGRA frame ordering. ([See docs](https://docs.opencv.org/3.1.0/d4/da8/group__imgcodecs.html#gabbc7ef1aa2edfaa87772f1202d67e0ce)]

**Comment by alalek** [comment]
Please try the latest code from "master" branch.

**Comment by tipsmiller** [comment]
I just pulled in 3.4.7 and was able to use the JavaCamera2View without this issue. The camera preview displays with the correct channels as expected.

**Comment by mshabunin** [comment]
Looks like it have been fixed in #14342 and/or #14505
